haystack/preview/components/generators/openai/chatgpt.py

```diff
@@ -0,0 +1,276 @@
+        streaming_callback = streaming_callback or self.streaming_callback
+        api_base_url = api_base_url or self.api_base_url
+
+        if system_prompt:
+            system_message = ChatMessage(content=system_prompt, role="system")
+        chats = []
+        for prompt in prompts:
+            message = ChatMessage(content=prompt, role="user")
+            if system_prompt:
+                chats.append([system_message, message])
+            else:
+                chats.append([message])
+
+        all_replies, all_metadata = [], []
+        for chat in chats:
+            completion = openai.ChatCompletion.create(
+                model=self.model_name,
+                api_key=self.api_key,
+                messages=[asdict(message) for message in chat],
+                stream=streaming_callback is not None,
+                **(self.model_parameters or model_parameters or {}),
+            )
+
+            replies: List[str]
+            metadata: List[Dict[str, Any]]
+            if streaming_callback:
+                replies_dict = {}
+                metadata_dict: Dict[str, Dict[str, Any]] = {}
+                for chunk in completion:
+                    chunk = streaming_callback(chunk)
+                    for choice in chunk.choices:
+                        if choice.index not in replies_dict:
+                            replies_dict[choice.index] = ""
+                            metadata_dict[choice.index] = {}
+
+                        if hasattr(choice.delta, "content"):
+                            replies_dict[choice.index] += choice.delta.content
+                        metadata_dict[choice.index] = {
+                            "model": chunk.model,
+                            "index": choice.index,
+                            "finish_reason": choice.finish_reason,
+                        }
+                all_replies.append(list(replies_dict.values()))
+                all_metadata.append(list(metadata_dict.values()))
+                self._check_truncated_answers(list(metadata_dict.values()))
+
+            else:
+                metadata = [
+                    {
+                        "model": completion.model,
+                        "index": choice.index,
+                        "finish_reason": choice.finish_reason,
+                        "usage": dict(completion.usage.items()),
+                    }
+                    for choice in completion.choices
+                ]
+                replies = [choice.message.content.strip() for choice in completion.choices]
+                all_replies.append(replies)
+                all_metadata.append(metadata)
+                self._check_truncated_answers(metadata)
+
+        return {"replies": all_replies, "metadata": all_metadata}
+
+    def _check_truncated_answers(self, metadata: List[Dict[str, Any]]):
+        """
+        Check the `finish_reason` the answers returned by OpenAI completions endpoint.
```

Comment:
```suggestion
        Check the `finish_reason` answers returned by OpenAI completions endpoint.
```
Suggesting instead of pushing because not sure if I'm right :)

Reply:
Uh that's quite messed up :see_no_evil: `answer` as well should go
```suggestion
        Check the `finish_reason` returned with the OpenAI completions.
```

Reply:
Ah okay that's much better, thank you! :D